### PR TITLE
Fix possible timing problem for C_Invoice allocation

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh/issues/7171

### CODE REVIEW

File: backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java [hunk 1]
```diff
@@ -86,24 +86,26 @@ public C_Invoice(@NonNull final PaymentReservationService paymentReservationServ
...
 		captureMoneyIfNeeded(invoice);
 		ensureUOMsAreNotNull(invoice);
 
 		C_Invoice_CreateExportData.scheduleOnTrxCommit(invoice);
 	}
 
+	private void autoAllocateAvailablePayments(final I_C_Invoice invoice)
+	{
+		allocationBL.autoAllocateAvailablePayments(invoice);
+		testAndMarkAsPaid(invoice);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java:106]: why do we have to call testAndMarkAsPaid again. we are already calling it at line 92....

I think we shall try calling it only once.

File: backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java [hunk 2]
```diff
@@ -236,6 +238,7 @@ private void allocateInvoiceAgainstCreditMemo(final I_C_Invoice creditMemo)
 			// Allocate the minimum between parent invoice open amt and what is left of the creditMemo's grand Total
 			invoiceBL.allocateCreditMemo(parentInvoice, creditMemo, amtToAllocate);
 		}
+		testAndMarkAsPaid(creditMemo);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java:241]: ahh, and here again

File: backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java [hunk 3]
```diff
@@ -293,20 +296,22 @@ private void linkInvoiceToPaymentIfNeeded(final I_C_Invoice invoice)
...
 
-	private void allocateInvoiceAgainstPaymentIfNeeded(final I_C_Invoice invoice)
+	private void allocateInvoiceAgainstPaymentIfNeeded(@NonNull final I_C_Invoice invoice)
 	{
 		final I_C_Order order = invoice.getC_Order();
 		if (paymentBL.canAllocateOrderPaymentToInvoice(order))
 		{
 			final I_C_Payment payment = order.getC_Payment();
 			allocationBL.autoAllocateSpecificPayment(invoice, payment, true);
+			testAndMarkAsPaid(invoice);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java:310]: and again

File: backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java [hunk 4]
```diff
@@ -293,20 +296,22 @@ private void linkInvoiceToPaymentIfNeeded(final I_C_Invoice invoice)
 			paymentDAO.save(payment);
 
 			allocationBL.autoAllocateSpecificPayment(invoice, payment, true);
+			testAndMarkAsPaid(invoice);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/interceptor/C_Invoice.java:299]: and again

**Comment by teosarca** [comment]
👎  because:

image

**Comment by TheBestPessimist** [comment]
Status update: Merged as it is, but added a big FIXME: [`09a235b` (#9876)](https://github.com/metasfresh/metasfresh/pull/9876/commits/09a235b8ae98532f2d8685c33107471e73523d5b)

**Pull request merged**